### DISCUSSION TITLE
Roadmap for 2022 Q2 in RoboticsAcademy development

### DISCUSSION THREAD

**Discussion by jmplaza** [JdeRobot/RoboticsAcademy#1691] [General]
Hi folks,

we have reached a pretty decent status in our RoboticsAcademy, although I see room for improvement in the next months. From my perspective, the five most important topics to improve are:

(1). More cute GUI
We recently refactored the frontend source code to deeper use of Django templates and resuse of widgets along several exercises. The next big step is to move towards REACT (@dmariaa has developed this [draft grid webpage](http://www.dmariaa.es:8000/react-web/), and @DanielinR will work on this)

(2). Optimization, speed up the execution
Here we are building a new exercise Python template (based on MultiProcessing instead of Threading, and so taking benefit of the multiple cores of the user's computer) (@iamarkaj , @RUFFY-369, @pcrecis ). In addition reviewing the GPU acceleration from Linux host machines and Windows host machines is worthy. 

(3). Robustness and incremental improvements:
Solve small bugs in several exercises. Introduce Ackermann cars in autonomous driving exercises (@Bytenaut ). Introduce LIDAR in some drone exercises.

(4). Adding new exercises.
We have three new upcoming exercises: one for visual based robot localization (@dvalladaresv), another for visual 3D odometry (@PabloAsensio), and another about Computer Vision Introduction using OpenCV (@felicidadaqm ). Beyond them, I think we should focus on improving the existing ones, instead of extending the current exercise collection. We recently added to more exercises: [laser-based localization with particle filter](https://www.youtube.com/watch?v=y7rBPpV2NdI) and [laser-mapping](https://www.youtube.com/watch?v=obHhJ-_Y96c) (@vladkrav )

(5). Moving to ROS2
We are working on the new RADI, based on ROS2 (@trunc8, @Carlosalpha1). Take a look at the "FollowPerson" drafts, both with [simulated TurtleBot2 robot](https://www.youtube.com/watch?v=fDAU465eVxQ) and with [physical TurtleBot2 robot](https://www.youtube.com/watch?v=54Jb4KJwyDM).

Some of the [projects proposed for GSoC-2022](https://jderobot.github.io/activities/gsoc/2022#ideas-list) are aligned with these topics.
What do you think? Further comments, ideas, PRs :-) ??

**Comment by jmplaza** [comment]
Regarding more fashionable GUI we may reuse some widgets from [robotwebtools](http://robotwebtools.org/) such as ros3djs, from [ROS board](https://github.com/dheera/rosboard) and from [WebViz](https://webviz.io/)

**Comment by pariaspe** [comment]
I would also like to suggest an improvement in user coding experience. 
Debbuging is always a pain in the neck, but it is even worse when using RoboticsAcademy since the text editor and pylint doesn't help you a lot.

**Reply by RUFFY-369** [reply]
I would like to point to a little improvement towards what @pariaspe  mentioned regarding user's coding experience that in the code editor all the lines after the `while True` are treated as iterative part so, even without an indention all the lines will keep functioning as part of the iteration which may confuse the user and is not quite pythonic which in case again makes debugging more confusing.

**Reply by RUFFY-369** [reply]
On a similar line UX, I would also suggest an improvement regarding the how gazebo is launched & viewed. Sometimes, its the case where there is no models to be seen in the gazebo GUI

**Comment by akshaynarisetti** [comment]
I would like to suggest an Improvement to the **Teleop Feature**

I think for anyone Teleop is to debug while running their code in the text editor but the current Architecture doesn’t support that i.e Running the code in the text editor while TeleOperation.

If we could improve that, something like debugging camera frames while trying out new cv algorithms while teleoperating around the environment would be possible.

Making it easy to Debug and Develop Computer Vision Applications(As a later upgrade we could also add a joystick 🕹 )

**Reply by pariaspe** [reply]
Hi @akshaynarisetti,
If I remember right we already have one in the FollowLine exercise. We are also thinking on adding it to more exercises, but we are currently refactoring the front-end, so we might wait until it is completed

**Reply by akshaynarisetti** [reply]
Hey @pariaspe, looks like the Tele-lop feature has to wait until the frontend is done.

I am working with Django on updating the card-views with on:hover video playing feature, considering React being implemented for it should I complete the feature?

**Reply by pariaspe** [reply]
Yes, as JM said in topic 1) we are also planning in introducing React to the frontend, so go ahead with it

**Reply by dmariaa** [reply]
I am moving the card views to react, so maybe I can do it there...

**Reply by akshaynarisetti** [reply]
Hi @dmariaa 
Could I also work along with you, I am very Interested.
